sip/sip-006-runtime-cost-assessment.md

```diff
@@ -746,3 +754,130 @@ types is defined as:
+in Clarity contracts. They must be written precisely as follows:
+
+1. The function should be `read-only`.
+2. The function should return a tuple with the keys `runtime`, `write_length`,
+`write_count`, `read_count`, and `read_length` (the execution cost measurement
+[categories](#measurements-for-execution-cost)).
+3. The values of the returned tuple should be Clarity expressions representing
+the asymptotic cost functions.
+3. The expression's constants and variables should be listed as arguments
+to the Clarity function.
+    * The arguments should be written in alphabetical order, and variables
+    should come before constants.
+
+Here is an example of a cost-assessment function where the runtime
+is set to `a * log(n) + b`, and all of the other parameters are constant.
+
+```Lisp
+(define-read-only (runtime_cost (n int) (a int) (b int))
+  {
+     runtime: (+ (* a (log n)) b)
+     write_length: a,
+     write_count: a,
+     read_count: a,
+     read_length: a,
+  })
+```
+
+### Making a Cost-Assessment Function Proposal
+
+Stacks holders can propose cost-assessment functions by calling the
+`propose-cost` function in the **Clarity Cost Voting Contract**. It
+can be called like so:
+
+```Lisp
+(contract-call?
+   .propose-cost
+   qualified-function-name
+   cost-function-address
+   cost-function-name)
+
+```
+
+This function will return a response containing the details of the proposal,
+and a proposal ID, if successful.
+
+Once submitted, a proposal is valid until either a supermajority miner vote
+is achieved, or until it expires, 14 days from when it was submitted.
+
+### Viewing Cost-Assessment Proposals
+
+To view pending cost-assessment function proposals, one can use the 
+following functions in the **Clarity Cost Voting Contract**:
+
+- `(get-proposal (id int))` returns a single proposal
+- `(get-proposals)` returns a list of all pending proposals
+- `(get-proposals-function (qualified-name string-ascii))` returns a list of pending proposals for a specific
+function
+
+### Voting for a Cost-Assessment Proposal
+
+**Miners**
+
+Miners can vote by creating a transaction that calls the cost assessment
+contract's `(vote (id int))` function and mining a block that includes this
+transaction. The `vote` function won't count the vote if the block wasn't
+mined by the node that signed that transaction. In other words, miners
+must **commit** their vote with their mining power.
+
+A super majority of miners must vote for a proposal in order for it to be
+accepted.
+
+**Stacks Holders**
+
+Stacks holders can signal their support for a cost-assessment function proposal
+by calling the `(signal (id int))` function in the **Clarity Cost
+Voting Contract**. These signals get weighted by the size of the signaler's
+STX holdings.
+
+Stacks holders signals help sway the miners one way or another on a proposal,
+however, they ultimately do not directly affect whether a proposal gets
+accepted by the miners.
```

Comment:
So, is this like a coin vote for STX holders?  Also, does this need to happen on-chain?  If the outcome is ultimately determined by miner votes alone, then maybe we (or the community) could implement an off-chain Blockstack app that lets you do coin-voting?

Reply:
I think the answer is yes here, but can you elaborate on what you mean by "coin voting"?

This on-chain voting idea came from @kantai. He suggested that miners would signal by including messages in their block's coinbase, and STX holders would signal via a voting contract. He mentioned that this protocol should probably look something like stacks holders are able to initiate cost upgrades while miners are be able veto them.

I suggested that perhaps miners could vote with the same contract that stacks holders use, as I think it'd be nice to have the votes all tallied in one place, so long as we're already using a contract.

Do you think that STX holders' signals should be counted as legitimate votes. It seems to me that they shouldn't have as much weight as the miners' votes, but perhaps they should be more than a signal...